src/internal.c

```diff
@@ -15277,71 +15277,88 @@ static WC_INLINE word32 UpdateHighwaterMark(word32 cur, word32 first,
 }
 #endif /* WOLFSSL_MULTICAST */
 
-/* diff must be already incremented by one */
-static void DtlsUpdateWindowGTSeq(word32 diff, word32* window)
+/* diff is the difference between the message sequence and the
+ * expected sequence number. 0 is special where it is an overflow. */
+static void _DtlsUpdateWindowGTSeq(word32 diff, word32* window)
 {
-    word32 idx, newDiff, temp, i;
+    word32 idx, temp, i;
     word32 oldWindow[WOLFSSL_DTLS_WINDOW_WORDS];
 
-    if (diff >= DTLS_SEQ_BITS)
+    if (diff == 0 || diff >= DTLS_SEQ_BITS)
         XMEMSET(window, 0, DTLS_SEQ_SZ);
     else {
         temp = 0;
         idx = diff / DTLS_WORD_BITS;
-        newDiff = diff % DTLS_WORD_BITS;
+        diff %= DTLS_WORD_BITS;
 
         XMEMCPY(oldWindow, window, sizeof(oldWindow));
 
         for (i = 0; i < WOLFSSL_DTLS_WINDOW_WORDS; i++) {
             if (i < idx)
                 window[i] = 0;
             else {
-                temp |= (oldWindow[i-idx] << newDiff);
+                temp |= (oldWindow[i-idx] << diff);
                 window[i] = temp;
-                temp = oldWindow[i-idx] >> (DTLS_WORD_BITS - newDiff - 1);
+                temp = oldWindow[i-idx] >> (DTLS_WORD_BITS - diff);
             }
         }
     }
     window[0] |= 1;
 }
 
-static WC_INLINE int _DtlsUpdateWindow(WOLFSSL* ssl, word16* next_hi,
-                                       word32* next_lo, word32 *window)
+int wolfSSL_DtlsUpdateWindow(word16 cur_hi, word32 cur_lo,
+        word16* next_hi, word32* next_lo, word32 *window)
 {
-    word32 cur_lo, diff;
+    word32 diff;
     int curLT;
-    word16 cur_hi;
-
-    cur_hi = ssl->keys.curSeq_hi;
-    cur_lo = ssl->keys.curSeq_lo;
 
     if (cur_hi == *next_hi) {
         curLT = cur_lo < *next_lo;
-        diff = curLT ? *next_lo - cur_lo - 1 : cur_lo - *next_lo + 1;
+        diff = curLT ? *next_lo - cur_lo : cur_lo - *next_lo;
     }
     else {
-        curLT = cur_hi < *next_hi;
-        diff = curLT ? cur_lo - *next_lo - 1 : *next_lo - cur_lo + 1;
+        if (cur_hi > *next_hi + 1) {
```

Comment:
why this and the check on 15330 are using `+ 1`? I think this can lead to incorrect results. Adding this line to the test:
```
     DUW_TEST(2, 0, next_hi, next_lo, window, 2, 4, 0, 0x19);
     DUW_TEST(99, 66, next_hi, next_lo, window, 99, 67, 0, 0x01);
 
+    DUW_TEST(100, 68, next_hi, next_lo, window, 100, 69, 0, 0x01);

```
indeed make the test fails.

Reply:
Sent a patch to @ejohnstown that fixes this and adds more testing.